### PR TITLE
[Death Knight] Venthyr Covenant

### PR DESCRIPTION
**Pull request by Armadk**
Implement Swarming Mist ability
Implement Impenetrable Gloom Conduit

Damage, target caps have both been tested on beta.

### CODE REVIEW

File: engine/class_modules/sc_death_knight.cpp [hunk 1]
```diff
@@ -9427,6 +9535,9 @@ void death_knight_t::init_gains()
   // Unholy
   gains.apocalypse                       = get_gain( "Apocalypse" );
   gains.festering_wound                  = get_gain( "Festering Wound" );
+
+  // Shadowlands / Covenants
+  gains.swarming_mist                    = get_gain( "swarming_mist" );
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: You don't need to tokenize the name here
```suggestion
  gains.swarming_mist                    = get_gain( "Swarming Mist" );
```

File: engine/class_modules/sc_death_knight.cpp [hunk 2]
```diff
@@ -6649,6 +6665,88 @@ struct summon_gargoyle_t : public death_knight_spell_t
...
+struct swarming_mist_buff_t : public buff_t
+{
+  swarming_mist_damage_t* damage; // (AOE) damage that ticks every second
+
+  swarming_mist_buff_t( death_knight_t* p ) :
+    buff_t( p, "swarming_mist", p -> covenant.swarming_mist ),
+    damage( new swarming_mist_damage_t( p ) )
+  {
+    cooldown -> duration = 0_ms; // Controlled by the action
+    set_refresh_behavior( buff_refresh_behavior::DURATION );
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: Is it even able to refresh? I wouldn't touch the default value unless necessary (though refresh behavior shouldn't ever be used anyway)

Comment (Armadk) [reply engine/class_modules/sc_death_knight.cpp]: Duration isn't long enough to refresh.  NGL it was just a straight copy of remorseless.  I had wondered about leaving that in.

File: engine/class_modules/sc_death_knight.cpp [hunk 3]
```diff
@@ -8646,11 +8746,15 @@ void death_knight_t::init_spells()
 
   // Shadowlands specific - Commented out = NYI
 
+  // Covenants
+  covenant.swarming_mist = find_covenant_spell( "Swarming Mist" );
+  covenant.swarming_mist_energize = covenant.swarming_mist->ok() ? find_spell( 312546 ) : spell_data_t::not_found();
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: If it's only used inside the swarming mist tick, you could just check it on the tick creation and store the value as a variable there.

Comment (Armadk) [reply engine/class_modules/sc_death_knight.cpp]:
```
orig
    swarming_mist              Count=   5.5| 60.000sec  DPE= 8292| 0.00%  DPET=7035  DPR=  0  pDPS=  0
    swarming_mist_damage       Count=  48.9|  5.904sec  DPE=  932|19.41%  DPET=  0  DPR=  0  pDPS=152  Miss= 0.00%  Hit=   800|   730|   905  Crit=  1598|  1460|  1811|16.59%
    swarming_mist                : start=  5.5 refresh=  0.0 interval= 60.0 trigger= 60.0 duration=  7.9 uptime= 14.54%

refactor
    swarming_mist              Count=   5.5| 60.000sec  DPE= 8315| 0.00%  DPET=7055  DPR=  0  pDPS=  0
    swarming_mist_damage       Count=  49.1|  5.872sec  DPE=  933|19.44%  DPET=  0  DPR=  0  pDPS=152  Miss= 0.00%  Hit=   800|   730|   905  Crit=  1599|  1460|  1811|16.69%
    swarming_mist                : start=  5.5 refresh=  0.0 interval= 60.0 trigger= 60.0 duration=  7.9 uptime= 14.52%
```

File: engine/class_modules/sc_death_knight.cpp [hunk 4]
```diff
@@ -6694,7 +6695,7 @@ struct swarming_mist_damage_t : public death_knight_spell_t
     if ( swarming_mist_energize_tick < swarming_mist_energize_target_cap )
     {
       p() -> resource_gain( RESOURCE_RUNIC_POWER,
-                 p() -> covenant.swarming_mist_energize->effectN( 1 ).resource( RESOURCE_RUNIC_POWER ),
+                 swarming_mist_energize->effectN( 1 ).resource( RESOURCE_RUNIC_POWER ),
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: I meant that you could just store the value from .resource as an int, not even the whole spelldata object

Comment (Armadk) [reply engine/class_modules/sc_death_knight.cpp]: that would.... make far too much sense.  I thought it odd from an efficiency standpoint to move the spelldata, but figured you wouldn't suggest it without reason.  Now it makes a lot more sense.

**Pull request merged**